### PR TITLE
Feat/conv 1 d

### PR DESCRIPTION
**Pull request by keshvituteja**
Added an implementation of the 1D convolution with stride 1.

### CODE REVIEW

**Comment by keshvituteja** [comment]
> I'm wondering, should this be called a convolution, or rather something like "stencil"? In the context of finite difference methods, convolution may be overly specific, but the same kernel is still useful.

We are planning to use it in context of Machine Learning (through pyGinkgo), so, we decided to name it 'Convolution'  for now. I/We are open to the naming scheme

**Comment by yhmtsai** [comment]
could you rebase your branch on origin/conv first to reduce the file changes?

**Comment by upsj** [comment]
The base commit seems incorrect, this should be based on/merged into `develop`

**Comment by yhmtsai** [comment]
@upsj It is not ready for merging into develop.
we will work on conv after this branch a little bit to check whether the interface/constructor makes sense or not.

File: core/matrix/conv.cpp [hunk 1]
```diff
@@ -50,7 +51,19 @@ template <typename ValueType>
...
+    const auto kernel_len = kernel_.get_size();
+
+    if (x_rows != b_rows + kernel_len - 1) {
+        throw DimensionMismatch(__FILE__, __LINE__, __func__, "x", x_rows, 1,
+                                "b + kernel - 1", b_rows + kernel_len - 1, 1,
+                                "x must have size = b + kernel - 1");
+    }
+
+
+    GKO_ASSERT_EQUAL_COLS(b, x);
```
Comment (yhmtsai) [inline core/matrix/conv.cpp:68]:
```suggestion
    GKO_ASSERT_EQUAL_COLS(b, x);
    GKO_ASSERT_EQUAL_COLS(b, dim<2>{1, 1});
```
does it only work on one single right hand side or can do several?

File: reference/matrix/conv_kernels.cpp [hunk 2]
```diff
@@ -24,10 +24,25 @@ namespace conv {
...
+          const array<ValueType>& kernel, const matrix::Dense<ValueType>* b,
+          matrix::Dense<ValueType>* x)
 {
-    GKO_NOT_IMPLEMENTED;
+    const auto b_size = b->get_size();                 // (N, 1)
+    const auto x_size = x->get_size();                 // (N + K - 1, 1)
+    const auto kernel_size = kernel.get_size();        // K
+    const auto* kernel_ptr = kernel.get_const_data();  // pointer to kernel data
+
+    for (int i = 0; i < x_size[0]; ++i) {
```
Comment (yhmtsai) [inline reference/matrix/conv_kernels.cpp]: it seems not to be correct.
your formula is x[0] = kernel[0] * b[0];
but it should be x[i] = sum_k kernel[k] * b[i+k], right?

Comment (yhmtsai) [inline reference/matrix/conv_kernels.cpp]: you need to use size_t to deal from the dense size

**Review by yhmtsai — CHANGES_REQUESTED** [review]

File: reference/matrix/conv_kernels.cpp (hunk 2 above)
Comment (keshvituteja) [reply reference/matrix/conv_kernels.cpp]: I updated it according to what torch does, it should work fine now [c629e5b](https://github.com/ginkgo-project/ginkgo/pull/1907/commits/c629e5bbd8e4c75916d97abb97533fd5b80666b2)

Comment (keshvituteja) [reply reference/matrix/conv_kernels.cpp]: I changed it to size_t in [630a24d](https://github.com/ginkgo-project/ginkgo/pull/1907/commits/630a24d7c03ec8495d441be9634ac63980ee303e), does size_t allow negative values? I am currently using `std::ptrdiff_t` for negative values

Comment (upsj) [reply reference/matrix/conv_kernels.cpp]: I tend to go by the rule of thumb: if you ever need to use a subtraction (even if you think the values will never become negative), you should use a signed type, so I would agree with using `int64` or `ptrdiff_t`

File: core/matrix/conv.cpp (hunk 1 above)
Comment (keshvituteja) [reply core/matrix/conv.cpp:68]: It works on single rhs only

File: core/matrix/conv.cpp [hunk 3]
```diff
@@ -50,7 +51,22 @@ template <typename ValueType>
...
+    using gko::detail::get_size;
+    const auto b_rows = get_size(b)[0];
+    const auto x_rows = get_size(x)[0];
+    const auto kernel_len = kernel_.get_size();
+
+    if (x_rows != b_rows + kernel_len - 1) {
+        throw DimensionMismatch(
+            __FILE__, __LINE__, __func__, "x", x_rows, 1,
+            "(b + 2*padding - kernel)/stride + 1",
+            (b_rows + 2 * 2 - kernel_len) / 1 + 1, 1,
```
Comment (yhmtsai) [inline core/matrix/conv.cpp]:
```suggestion
            (b_rows + 2 * 0 - kernel_len) / 1 + 1, 1,
```
shouldn't padding be zero now?

File: reference/matrix/conv_kernels.cpp [hunk 4]
```diff
@@ -24,10 +24,34 @@ namespace conv {
...
-    GKO_NOT_IMPLEMENTED;
+    const auto b_size = b->get_size();                 // (N, 1)
+    const auto x_size = x->get_size();                 // (N + K - 1, 1)
+    const auto kernel_size = kernel.get_size();        // K
+    const auto* kernel_ptr = kernel.get_const_data();  // pointer to kernel data
+    int stride = 1;
+    int padding = 2;
+    int output_length = (x_size[0] + 2 * padding - kernel_size) / stride + 1;
+
+    for (gko::size_type i = 0; i < x_size[0]; ++i) {
```
Comment (yhmtsai) [inline reference/matrix/conv_kernels.cpp:38]: because you are using substraction later, so I will say you can use gko::int64 directly as Tobias mentioned

File: core/matrix/conv.cpp (hunk 3 above)
Comment (yhmtsai) [reply core/matrix/conv.cpp]: padding equal 2 in your example such that it become numpy result.
IMO, we should follow the default from torch not numpy.
also, if it needs to fit numpy, it should be something like ((x-1)*stride + kernel - b)/2

File: reference/matrix/conv_kernels.cpp (hunk 4 above)
Comment (yhmtsai) [reply reference/matrix/conv_kernels.cpp:38]: also I think start with output will be easier than with input

File: reference/matrix/conv_kernels.cpp [hunk 5]
```diff
@@ -24,10 +24,34 @@ namespace conv {
...
+        ValueType sum = zero<ValueType>();
+        std::ptrdiff_t start =
+            static_cast<std::ptrdiff_t>(i * stride) - padding;
+        for (gko::size_type j = 0; j < kernel_size; ++j) {
+            std::ptrdiff_t b_idx =
+                start +
+                static_cast<std::ptrdiff_t>(
+                    j);  // calculate the index in b's row based on the current
```
Comment (yhmtsai) [inline reference/matrix/conv_kernels.cpp]:
```suggestion
                static_cast<std::ptrdiff_t>(
                    j) * stride;  // calculate the index in b's row based on the current
```

**Review by yhmtsai — CHANGES_REQUESTED** [review]

File: core/matrix/conv.cpp (hunk 3 above)
Comment (keshvituteja) [reply core/matrix/conv.cpp]: The example also matches the torch result. We can change the padding inside the `conv_kernels.cpp`. It's result would be same as this function:
`torch.nn.functional.conv1d(x, w, bias=None, stride=1, padding=2)`

We can accept padding in arguments but currently we don't do that, so i hard coded it to be 2. The function also works for zero-padding. For zero padding, it should be similar to:
``torch.nn.functional.conv1d(x, w, bias=None, stride=1, padding=0)``

File: reference/matrix/conv_kernels.cpp (hunk 5 above)
Comment (keshvituteja) [reply reference/matrix/conv_kernels.cpp]: We might not need the multplication with stride here since we already multiply with stride while calculating`start` in line 41:
`gko::int64 start =static_cast<gko::int64>(i * stride) - padding;`

File: core/matrix/conv.cpp (hunk 3 above)
Comment (keshvituteja) [reply core/matrix/conv.cpp]: I changed it to zero but it will work if change it to any other padding as well

File: reference/matrix/conv_kernels.cpp (hunk 5 above)
Comment (yhmtsai) [reply reference/matrix/conv_kernels.cpp]: I see. I mix dilation and stride here. it is needed in dilation not stride

File: reference/matrix/conv_kernels.cpp [hunk 6]
```diff
@@ -24,10 +24,33 @@ namespace conv {
...
+        gko::int64 start = static_cast<gko::int64>(i * stride) - padding;
+        for (gko::size_type j = 0; j < kernel_size; ++j) {
+            gko::int64 b_idx =
+                start +
+                static_cast<gko::int64>(
+                    j);  // calculate the index in b's row based on the current
+                         // position in x and the kernel's stride and padding
+            if (b_idx >= 0 && b_idx < static_cast<gko::int64>(b_size[0])) {
+                sum += kernel_ptr[j] * b->at(static_cast<gko::size_type>(b_idx),
+                                             0);  // direct pointer access
```
Comment (yhmtsai) [inline reference/matrix/conv_kernels.cpp:49]: you do not need to do it because no subtraction here.
